src/main/java/ru/ralsei/whatcanyousee/logic/MazeGameMap.java

```diff
@@ -0,0 +1,809 @@
+     *
+     * Basically, to create a toogle one just has to implement that special event.
+     */
+    protected abstract class Toogle {
+        /**
+         * Special event cause by pressing this toogle.
+         */
+        protected abstract void use();
+    }
+
+    /**
+     * List of all monster in the game.
+     */
+    private final ArrayList<Monster> monsters = new ArrayList<>();
+
+    /**
+     * Adds new monster to the list.
+     */
+    protected void addMonster(Monster monster) {
+        monsters.add(monster);
+    }
+
+    public ArrayList<Monster> getMonsters() {
+        return monsters;
+    }
+
+    /**
+     * Class representing the monster in maze.
+     *
+     * Maze
+     */
+    protected abstract class Monster {
+        /**
+         * Image of this monster.
+         */
+        private int imageId;
+
+        /**
+         * Initial position in the maze.
+         */
+        private int initialX;
+        private int initialY;
+
+        /**
+         * Current position in the maze.
+         */
+        private int currentX;
+        private int currentY;
+
+        /**
+         * True if monster is ready to move -- he will be moved on the next tick.
+         */
+        abstract protected boolean readyToMove();
+
+        /**
+         * Moves monster to the given cell.
+         */
+        void moveTo(Cell cell) {
+            this.getCurrentCell().numberOfMonsters--;
+            currentX = cell.getX();
+            currentY = cell.getY();
+            cell.numberOfMonsters++;
+        }
+
+        /**
+         * Updates monster's state on gameplay tick -- for instance, monster could be moving each
+         * 5 ticks, so here he updates his counter.
+         */
+        abstract protected void updateOnTick();
+
+        int getImageId() {
+            return this.imageId;
+        }
+
+        protected void setImageId(int imageId) {
+            this.imageId = imageId;
+        }
+
+        protected int getInitialX() {
+            return initialX;
+        }
+
+        protected void setInitialX(int initialX) {
+            this.initialX = initialX;
+            currentX = initialX;
+        }
+
+        protected int getInitialY() {
+            return initialY;
+        }
+
+        protected void setInitialY(int initialY) {
+            this.initialY = initialY;
+            currentY = initialY;
+        }
+
+        Coordinates getCurrentCoordinates() {
+           return new Coordinates(currentX, currentY);
+        }
+
+        protected Cell getCurrentCell() {
+            return getCell(getCurrentCoordinates());
+        }
+
+        /**
+         * Monster tries to kill the player -- usually if he stepped on the cell with the player.
+         */
+        abstract protected boolean tryToKill();
+
+        /**
+         * Decreasing tickTo, but if it's 0, sets it to tickPer.
+         */
+        protected int decreaseTick(int tickTo, int tickPer) {
```

Comment:
Can be made static

Reply:
However, it cannot, "inner classes cannot have static declarations". I cannot move this method outside, because it has meaning only in Monster class, and It's a bad idea (IMHO) to make Monster class static.

Reply:
The official [tutorial ](https://docs.oracle.com/javase/tutorial/java/javaOO/whentouse.html) says

> Use a non-static nested class (or inner class) if you require access to an enclosing instance's non-public fields and methods. Use a static nested class if you don't require this access.

And it seems exactly like my case: Monster's extended classes use Map's protected methods such as setMessageLost(), setPlayerWon(), etc. So in order to make Monster class static, I have to store its MazeGameMap in a field (it uses some methods of MazeGameMap), but in order to use this field in the extended classes, they should be non-static, therefore does not need this field, therefore it becomes wtf code.